Liquibase 4.3.0 throws NullPointerExceptions without snakeyaml in the classpath

## Description
Snakeyaml is the 3rd party library we use for JSON and YAML parsing. 

It has always been an optional component, traditionally only needed for json/yaml changelogs and offline snapshots (which are json-formatted).

As we are adding hub.liquibase.com based functionality, that also requires snakeyaml for the json rest API calls.

With 4.3.0, there is an issue where the "get the HubService" implementation call returns null if we cannot instantiate the standard hubservice class, and then the code that expects a service to be returned throws the null pointer exception.

## Impact
This should NOT impact:

* CLI users (because snakeyaml is included in the CLI distribution)
* Liquibase-Maven-Plugin users, because Maven appears to include snakeyaml in the liquibase classpath with the `<optional>true</optional>` setting we have in the plugin's pom.xml
* Any spring boot or other embedded users, where snakeyaml is not explicitly included as a dependency

This SHOULD impact:

* Any spring boot or other embedded setups, where snakeyaml is not explicitly included as a dependency. We have it listed as <optional>true</option> in our pom.xml, so it should not be automatically included
* Ant users
* Anyone else running liquibase without snakeyaml.jar in their classpath.

## Resolution Options
1. We can make snakeyaml a required dependency. As we continue to add hub-based functionality, the ability to use json will become more and more central that it's not worth keeping it as an optional dependency.
1. Continue to keep snakeyaml optional, and improve test coverage around how to handle cases where the library is not available

NOTE: Using a different library than snakeyaml for json parsing is a different issue.

## Liquibase Internal QA Test Requirements
* Verify Liquibase CLI update does not throw a null pointer if snakeyaml is not in path.
* Verify Liquibase Ant update does not throw a null pointer if snakeyaml is not in the path.
* Verify Liquibase Servlet update does not throw a null pointer if snakeyaml is not in the path.
* Verify Liquibase Maven update does not throw a null pointer.
  * Fix should have no impact on Maven for reasons described in the Impact section.

* Verify all automated functional CLI tests pass on Jenkins.
##### Jenkins is an internal build server and results are not visible to the wider community.




┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-1212) by [Unito](https://www.unito.io/learn-more)
┆Fix Versions: Community 4.3.1,Liquibase 4.3.1

I also encountered the issue with 4.3.0.
Here is the stacktrace for reference:
```
liquibase.exception.LiquibaseException: java.lang.NullPointerException                             
        at liquibase.Liquibase.runInScope(Liquibase.java:2348)                                     
        at liquibase.Liquibase.update(Liquibase.java:216)                                          
        at liquibase.Liquibase.update(Liquibase.java:202)                                          
        at liquibase.integration.commandline.Main.doMigration(Main.java:1764)                      
        at liquibase.integration.commandline.Main$1.lambda$run$0(Main.java:364)                    
        at liquibase.integration.commandline.Main$1$$Lambda$3.0000000010621090.run(Unknown Source) 
        at liquibase.Scope.lambda$child$0(Scope.java:160)                                          
        at liquibase.Scope$$Lambda$4.0000000010CACB20.run(Unknown Source)                          
        at liquibase.Scope.child(Scope.java:169)                                                   
        at liquibase.Scope.child(Scope.java:159)                                                   
        at liquibase.Scope.child(Scope.java:138)                                                   
        at liquibase.integration.commandline.Main$1.run(Main.java:363)                             
        at liquibase.integration.commandline.Main$1.run(Main.java:193)                             
        at liquibase.Scope.child(Scope.java:169)                                                   
        at liquibase.Scope.child(Scope.java:145)                                                   
        at liquibase.integration.commandline.Main.run(Main.java:193)                               
        at liquibase.integration.commandline.Main.main(Main.java:156)                              
Caused by: java.lang.NullPointerException                                                          
        at liquibase.hub.HubUpdater.register(HubUpdater.java:306)                                  
        at liquibase.Liquibase.lambda$update$1(Liquibase.java:242)                                 
        at liquibase.Liquibase$$Lambda$8.0000000010CA73E0.run(Unknown Source)                      
        at liquibase.Scope.lambda$child$0(Scope.java:160)                                          
        at liquibase.Scope$$Lambda$4.0000000010CACB20.run(Unknown Source)                          
        at liquibase.Scope.child(Scope.java:169)                                                   
        at liquibase.Scope.child(Scope.java:159)                                                   
        at liquibase.Scope.child(Scope.java:138)                                                   
        at liquibase.Liquibase.runInScope(Liquibase.java:2343)                                     
        ... 16 more
```

--- a/liquibase-core/src/main/java/liquibase/hub/HubServiceFactory.java
+++ b/liquibase-core/src/main/java/liquibase/hub/HubServiceFactory.java
@@ -1,10 +1,17 @@
 package liquibase.hub;
 
+import liquibase.changelog.RanChangeSet;
+import liquibase.exception.LiquibaseException;
+import liquibase.hub.model.*;
 import liquibase.plugin.AbstractPluginFactory;
+
+import java.util.List;
+import java.util.UUID;
 
 public class HubServiceFactory extends AbstractPluginFactory<HubService> {
 
     private String offlineReason;
+    private static final FallbackHubService fallbackHubService = new FallbackHubService();
 
     protected HubServiceFactory() {
     }
@@ -20,7 +27,11 @@ public class HubServiceFactory extends AbstractPluginFactory<HubService> {
     }
 
     public HubService getService() {
-        return getPlugin();
+        final HubService plugin = getPlugin();
+        if (plugin == null) {
+            return fallbackHubService;
+        }
+        return plugin;
     }
 
     public boolean isOnline() {
@@ -33,5 +44,112 @@ public class HubServiceFactory extends AbstractPluginFactory<HubService> {
 
     public void setOfflineReason(String offlineReason) {
         this.offlineReason = offlineReason;
+    }
+
+    /**
+     * If StandardHubService can't be instantiated for some reason (like a missing snakeyaml), return this implementation
+     * to avoid null pointer exceptions.
+     *
+     * Once StandardHubService can always be instantiated, this fallback can be removed
+     */
+    private static class FallbackHubService implements HubService {
+
+        @Override
+        public int getPriority() {
+            return PRIORITY_NOT_APPLICABLE;
+        }
+
+        /**
+         * Always offline
+         */
+        @Override
+        public boolean isOnline() {
+            return false;
+        }
+
+        @Override
+        public HubUser getMe() throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public Organization getOrganization() throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public Project getProject(UUID projectId) throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public List<Project> getProjects() throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public Project createProject(Project project) throws LiquibaseException {
+            return null;
+        }
+
+        @Override
+        public void setRanChangeSets(Connection connectionId, List<RanChangeSet> ranChangeSets) throws LiquibaseHubException {
+
+        }
+
+        @Override
+        public Connection getConnection(Connection exampleConnection, boolean createIfNotExists) throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public List<Connection> getConnections(Connection exampleConnection) throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public HubRegisterResponse register(String email) throws LiquibaseException {
+            return null;
+        }
+
+        @Override
+        public Connection createConnection(Connection connection) throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public HubChangeLog createChangeLog(HubChangeLog hubChangeLog) throws LiquibaseException {
+            return null;
+        }
+
+        @Override
+        public HubChangeLog getHubChangeLog(UUID changeLogId) throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public Operation createOperation(String operationType, HubChangeLog changeLog, Connection connection) throws LiquibaseHubException {
+            return null;
+        }
+
+        @Override
+        public OperationEvent sendOperationEvent(Operation operation, OperationEvent operationEvent) throws LiquibaseException {
+            return null;
+        }
+
+        @Override
+        public String shortenLink(String url) throws LiquibaseException {
+            return null;
+        }
+
+        @Override
+        public void sendOperationChangeEvent(OperationChangeEvent operationChangeEvent) throws LiquibaseException {
+
+        }
+
+        @Override
+        public void sendOperationChanges(OperationChange operationChange) throws LiquibaseHubException {
+
+        }
     }
 }

